### ISSUE TITLE
subpixel.py API Standardization

### ISSUE THREAD

**Issue by jlaura** [USGS-Astrogeology/autocnet#553]
subpixel.py needs some serious love. The code is basically untestable in its current form because the functions are so tightly coupled to the NCG via database sessions. Most of the functions are (1) connecting the the DB, (2) running queries to get objects to operate on, (3) operating on the objects and (4) writing updated results back to the DB.

Ideally, these functions would only operate on the objects (or dict representations of the objects. We still require the other capabilities - they just need to be modularized into something testable. Additionally, this will allow any other caller to make use of our subpixel registration algorithms. The way that subpixel.py is setup is also impacting ground.py as the APIs (and structures) are largely consistent.

Secondly, we have a number of matchers and transformation techniques that all perform various data reads, transformations, and applications of matching functions. These need to be a single matching function or interface (class?) that can be parametrized to support:

- arbitrary matching function that takes in two numpy arrays plus any parametrizing kwargs and  returns either offsets or updated coordinates
- ability to apply arbitrary transformations to the aforementioned numpy arrays before passing to the matchers
- ability to apply the diverse transformation
- ability to work in the correct dtype space (though these functions should never explicitly know about dtypes)

## Callers
Who are the callers of these functions?
  - NCG directly via syntax sugar. These functions run in serial.
  - can_submit via the apply method on the NCG. These funcs run in parallel via many NCG instances.
  - third party callers (e.g., jupyter notebooks for change detection)

## Current funcs in subpixel.py 

| func | logic | returns|
|-----|------|--------|
|subpixel_phase|clips ROIs and runs correlator | dx, dy, error, None | 
|subpixel_transformed_template| clips ROIs (checking dtypes), apply a passed in transformation, apply the correlator, blackout the transformation | dx, dy, error, corrmap |
| subpixel_template_classic|  clips ROIs and runs correlator | dx, dy, error, corrmap | 
| subpixel_template | clips ROIs (checking dtypes), runs the correlator, checks that correlation is not near an edge | dx, dy, error, corrmap | 
| subpixel_ciratefi | clips the ROIs and runs the correlator | dx, dy, error |
| iterative_phase | dispatches ROI work to subpixel_phase, iteratively applies correlator | dx, dy, metrics |
| estimate_affine_transformation | estimates a transform using src and dest coords | transformation obj|
| geom_match_simple | pull coordinates for transformation estimation, estimation of the transformation, logic to read the data sets with proper dtype, apply transform, apply a match func | sample, line, dist, metric, temp_corrmap |
| geom_match_classic | pull coordinates for transformation estimation, estimation of the transformation, logic to read the data sets with proper dtype, apply transform, apply a match func | sample, line, dist, metric, temp_corrmap |
| geom_match | pull coordinates for transformation estimation, dispatch estimation of transform, call subpixel_transformed_template to handle clipping and matching | x, y, dist, metric, corrmap 
| subpixel_register_measure | create a db session, get the reference measure / measure to operate on, apply geom_match_simple, check should update, update db object | resultlog dict |
| subpixel_register_point | create DB session, get point/measures to register, apply a matcher, check if should update, update db objs | resultlog dict |
| subpixel_register_points | open a db session to get a list of point ids, call subpixel_register_point | None | 

## Proposed changes
- The dtype checking needs to occur on the ROI object so that it is standardized in a single place.
- The byte scaling of data needs to be an argument on clip or a method on the ROI.
- Move estimate_affine_transformation out of subpixel.py and into transformation. This could go onto the ROI object or into a new file. Not 100% sure.
- subpixel_register_* should probably be in a separate file. These are fundamentally different algorithms than the matchers.
- subpixel_register_* should take in Point/Measure objs (or dicts) and return updated Point/Measure objs (or dicts)

## Questions:
- Is it better for subpixel_register_* to use dicts or the models? In the case of the former, we have standard python types that any caller can use without having to import our ORM definitions. This should make these methods more usable with the CG object that has not DB connection. This does mean that any caller using the DB (NCG, acn_submit) will be responsible for taking dicts and converting into DB objs (which is trivial).
- Should these matchers be a subpixel class that has 2+ ROI objects. ROI objects take care of clipping, transformations, dtype reads. The subpixel class takes care of apply an arbitrary matcher. Who then takes care of inverting the transformation to get real coordinates? How else might we design the matcher API?

## Longer term question to maybe address here?
- Can we get the transformations properly occurring on the ROI objs (and not on the whole image)? The transformation is a major, non-trivial time cost when using.

**Comment by Kelvinrr** [comment]
Some thoughts on how to handle registration funcs after writing the fourier mellen matcher: 

* Make the "source, destination" names something more obvious. Sometimes, names like source/destination doesn't make it super clear which image is being moved to match which. I like [skimage's names for phase_cross_correlation's](https://scikit-image.org/docs/stable/api/skimage.registration.html#phase-cross-correlation) of `reference_image, moving_image` which feels super obvious which is which. 
* I think we should simplify the inputs and not use `sx, sy, dx, dy` type inputs and just use the centers so the moving image is always moved to best match the reference image from their centers. 
* Use affine matrices or objects for returning registration results. Whether it's geom_match, fourier_mellen, or a subpixel template, we are usually trying to solve for more than just translation. This is with the caveat that we shop for a difference affine lib for transforming images than skimage since I find them obnoxious to work with.

**Comment by jlaura** [comment]
> This is with the caveat that we shop for a different affine lib for transforming images than skimage since I find them obnoxious to work with.

Some options:
- [spicy.ndimage.affine_transform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html) - this does not do the estimation, but does apply the transformation.
- [affine6p](https://gitlab.com/yoshimoto/affine6p-py) - pure python estimation library. Very 'simple' in terms of both API and a standard implementation.
- OpenCV2 - this is one of those really nice to have, but a pain to manage dependencies of the library
- [Pillow-affine](https://pillow-affine.readthedocs.io/en/latest/pillow_affine.html) - provides a convenience interface to pillow.transform. No estimation here, but access to component parts of the transformation.
- others?

**Comment by Kelvinrr** [comment]
> So like center_reference_image = (5,5) and center_moving_image = (15,15)?

no, more like the registration functions in subpixel.py assume that you have already "clipped" the images in ROIs and internally always use `center_x,center_y = ref_image.shape/2` and ensure both reference and moving images are the same shape. It seems some errors come from having to select two window sizes, one for ROI and one for the matcher.

**Comment by jlaura** [comment]
@Kelvinrr I updated the initial post with your suggestion. 

@acpaquette @arsanders if interested, do you all want to weigh in on this issue since you have recently added matchers? I would like to have someone work on the standardization this FY, so the more eyes and input, the better shot we have of having a cleaner implementation planned.

**Comment by jlaura** [comment]
I have had some time doing more work related tangentially to this issue and want to get some more thoughts down as they relate to the initial post.

Here are the initial proposed changes that I am still in favor of:

- The dtype checking needs to occur on the ROI object so that it is standardized in a single place.
- The byte scaling of data needs to be an argument on clip or a method on the ROI.
- Move estimate_affine_transformation out of subpixel.py and into transformation. This could go onto the ROI object or into a new file. Not 100% sure.
- subpixel_register_* should probably be in a separate file. These are fundamentally different algorithms than the matchers.
- subpixel_register_* should take in Point/Measure objs (or dicts) and return updated Point/Measure objs (or dicts)
- registration functions accept ndarrays as inputs and implicitly assume that the center of those ndarrays are the a priori match (credit @Kelvinrr)
 
Here are some new proposals as well:
- Separate geom_func and match func completely. The geom_func should return the affine transformed arrays or an ROI with the ability to extract an affine transformed subset. The geom funcs should also return the affine transformation object. The match funcs should operate on any arbitrary ROIs (ndarrays or GeoDatasets). 
- The funcs that want to sub pixel register a point / measure should handle calling the geom_func, dealing with the results and any preprocessing requirements (e.g., application of a filter). The registration func can then pass the arrays to the matcher, get the results and do whatever it likes (e.g., apply the inverse affine transform). Something abstract like:

```python
def subpixel_register_point(dict_data_structure, geom_args, matcher_args):
    get the reference measure and setup the roc
    for each other measure:
        get the to match roi
        match_to_arr, match_from_arr, affine_transform = apply the geom matcher(reference_roi, to_match_roi, genom_args)
        

        for each set of args in match args:
            result = matcher(match_to_arr, match_from_arr, args)
            transform result with affine
            build an entry for the return obj that is in the same format as the input dict (see below for what we might pass)
            append transformed result to return obj
        
        return
```

Then it is possible to call the above directly and do something with the result or write a wrapper that calls the above and operates on the result dict

- (Quite Opinionated) Get the ncg out of the point registration. Right now `subpixel_register_point` uses the NCG to get the point/measures. This is coupling subpixel register point to the database / database as modeled by the ORM. Instead, just pass in a data structure (dict) and return a data structure of updates. Then encapsulate any writing into the caller (honestly all the way up into the cluster_submit caller for batch runs. Then a jupyter notebook user could call these funcs, analyze the results as desired, and do whatever. All distinct from the DB. 

```python
# pass the dict representation of the model

{'point': {'id': 1, 
               'pointIgnore':False, 
               'measures' [
                    {'id':1, etc...},
                    {'id':2, etc...}
                ], 
                'next_attr': foo,
                etc...
}
```

- (Even more opinionated) Remove all the matchers except classic / simple. Everything else is questionable quality right now.

**Comment by jlaura** [comment]
In the dev branch, we have two different estimate_affine_transformation funcs with different arguments. This is a natural place to start refactoring I suppose. My plan to start tackling this is:

1. Fix the estimate_affine_transformation duplication by refactoring that out into the transformation module.
2. Attempt to get an integration test written (using a cropped CTX pair) that will run the full matching procedure. As I work on that, I'll figure out where/how to mock the different components. Once we have the integration test, I think we can start to refactor with less fear of introducing a breaking change.

**Comment by Kelvinrr** [comment]
@jlaura What the DOD for this issue? After a new Api is implemented or simple after we have a concrete plan for what needs to change?

**Comment by jlaura** [comment]
New API is implemented. In this sprint, having a plan is 'done'.

**Comment by jlaura** [comment]
Move to a common function signature: `def func(reference_roi, moving_roi, affine, **kwargs):`

## Update subpixel_items:
In priority order:
- [ ] subpixel_template_classic
- [ ] mutual_information
- [ ] iterative_phase
- [ ] subpixel_ciratefi (split to components; low priority)

## Update geom_match:
The output affine from `geom_match` is used as an optional input in the above, refactored funcs.

```python
def estimate_affine_from_sensors(referenc_image, moving_image) -> AffineTransformation:
```

## Moving the funcs for a consistent API
What is the file architecture of these? Let's update first and then we can move them into different modules.

## Usage by a user
A caller would: (1) estimate an affine if desired, (2) call one or more subpixel_* method sand get affine transformations back (this function does what it needs to do with the ROI extents and the passed affine), (3) apply the affine to the a priori coordinates to get an updated position.
